resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeHelper.scala

```diff
@@ -0,0 +1,121 @@
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.spark.deploy.yarn
+
+import java.lang.{Integer => JInteger, Long => JLong}
+import java.lang.reflect.InvocationTargetException
+
+import scala.collection.mutable
+import scala.util.control.NonFatal
+
+import org.apache.hadoop.yarn.api.records.Resource
+
+import org.apache.spark.internal.Logging
+import org.apache.spark.util.Utils
+
+/**
+ * This helper class uses some of Hadoop 3 methods from the YARN API,
+ * so we need to use reflection to avoid compile error when building against Hadoop 2.x
+ */
+private object ResourceTypeHelper extends Logging {
+  private val AMOUNT_AND_UNIT_REGEX = "([0-9]+)([A-Za-z]*)".r
+  private val RESOURCE_TYPES_NOT_AVAILABLE_ERROR_MESSAGE =
+    "Ignoring updating resource with resource types because " +
+        "the version of YARN does not support it!"
+
+  def setResourceInfoFromResourceTypes(
+      resourceTypes: Map[String, String],
+      resource: Resource): Resource = {
+    require(resource != null, "Resource parameter should not be null!")
+
+    logDebug(s"Custom resource types requested: $resourceTypes")
+    if (!ResourceTypeHelper.isYarnResourceTypesAvailable()) {
+      if (resourceTypes.nonEmpty) {
+        logWarning(RESOURCE_TYPES_NOT_AVAILABLE_ERROR_MESSAGE)
+      }
+      return resource
+    }
+
+    val resInfoClass = Utils.classForName(
+      "org.apache.hadoop.yarn.api.records.ResourceInformation")
+    val setResourceInformationMethod =
+      resource.getClass.getMethod("setResourceInformation", classOf[String],
+        resInfoClass)
+    resourceTypes.foreach { case (name, rawAmount) =>
+      try {
+        val AMOUNT_AND_UNIT_REGEX(amountPart, unitPart) = rawAmount
+        val amount = amountPart.toLong
+        val unit = unitPart match {
+          case "g" => "G"
+          case "t" => "T"
+          case "p" => "P"
+          case _ => unitPart
+        }
+        logDebug(s"Registering resource with name: $name, amount: $amount, unit: $unit")
+        val resourceInformation =
+          createResourceInformation(name, amount, unit, resInfoClass)
+        setResourceInformationMethod.invoke(resource, name,
+          resourceInformation.asInstanceOf[AnyRef])
+      } catch {
+        case _: MatchError => throw new IllegalArgumentException(
+          s"Resource request for '$name' ('$rawAmount') " +
+              s"does not match pattern $AMOUNT_AND_UNIT_REGEX.")
+        case e: InvocationTargetException =>
```

Comment:
I suggest: `case e: InvocationTargetException if e.getCause != null =>` and then you don't need the `throw e` case.

Reply:
Good point, fixed